### ISSUE TITLE
rework reviewer.ts

### ISSUE THREAD

**Issue by dae** [ankitects/anki#1120]
reviewer.ts could do with an update at one point, and it would be best if we can make any changes to it in one go, instead of trickling out breaking changes over time. This issue is for discussing and tracking what needs to be done.

Some things that should probably be addressed:

- [ ] as mentioned on #1119, we should probably remove the opacity changes and the promise race
- [ ] moving away from Python style _underscorePrefixed() functions, and instead explicitly exporting our public API. We could use an approach like editor/index_wrapper.ts, or maybe we should consider exporting them into a single global, like `anki` or `anki.reviewer`?
- [ ] _updateQA() and the hook system could use a rethink - we no longer need to fade in and out, and the current mix of promises and hooks is a bit tricky to follow - could it be simplified perhaps? onShownHook and onUpdateHook were exposed to allow templates to perform different actions before the content was revealed and after it was revealed, but without an opacity change, perhaps one of them is redundant?
- [ ] should we perhaps pass the MathJax and image promises to hooks in an object, so they can call .then() on them if they want to?
- [ ] @glutanimate has requested a permanent hook that doesn't get reset on each card load. The behaviour can be emulated by injecting the same hook each load via the Python API, but perhaps it's worth considering adding a separate long-lived one as well, which would get spread into the per-card one.

Thoughts? Any other things that could do with a change?

**Comment by hgiesel** [comment]
1. If we just remove the opacity changes, simple MathJax like `\( \begin{array}{cc} a & b \\ c & d \end{array} \)` will already cause a flash of unrendered content for me.
I personally also have a horse in this game, as my add-on Closet also heavily builds on `onUpdatePromise`, and it operating on hidden content.

2. `onUpdatePromise` is useful for changing the content before it is shown, for anything template related. `onShownPromise` can be useful for showing any of animation on card shown, start a timer, etc.. I think both hooks still have their reason for existing.

**Comment by hgiesel** [comment]
One way to get rid of the flash for most of the cards is, if we had some logic to *test* whether MathJax was used on the card, and only hide if they do. That's actually similar to how AnkiDroid does it.

**Comment by dae** [comment]
Ah, scratch that idea then.

It would be great if we could do the rendering in the background, then atomically swap the old content with the new. That hasn't proved practical in the past due to concerns with the same id being used in the old and new content, and preloading too early resulting into html audio possibly autoplaying, but maybe it would be achievable if we used a shadow dom and preloaded right before updating the display?

**Comment by hgiesel** [comment]
The issue with _swapping HTML_ (e.g. using `Node.prototype.cloneNode` or `innerHTML`)  would also be (other than IDs) , that we loose any event listeners in that chunk, which will be very disruptive for templates.

What we could do however is render the new card, displaying it offscreen (e.g. `position: absolute; left: -99999rem`), and once it's done, we dismount the old card, and move the new one in view. Hopefully this would get rid of the flash (however it would be very disappointing if it didn't ^^;)

A shadow DOM would be a good solution for the ID problem, but it will also be very disruptive, as users could not use common DOM functions like `document.querySelector` anymore, add CSS to the `<head>` for activating it, or style with `.nightMode .myClass`.
If we went down this route, we certainly would have to make it opt-in or only for new templates like you said in the other PR, and possibly offer users a bit of an API to work with.

Another issue is the audio for sure. Is there any strong reason why the audio is handled through Python right now? Playing around with [&lt;audio>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/audio) and inserting it to the reviewer seems to work quite well.

**Comment by dae** [comment]
WebEngine (and WebKit before it) are provided with limited codec support due to licensing issues, and the support has been buggy in the past. If we switched to web audio now, we'd likely prevent certain file formats from playing, or would need to maintain both the old and new audio playing codepaths in parallel, and coordinating sequential autoplay across audio, video and TTS would be a pain.

Presumably we could find solutions for .nightMode and css injection, but it does sound like a bigger task than I was hoping for at the moment. Maybe the best we can do for now is skipping the opacity change if no mathjax+no extra update hooks are found, like you suggested?

Will add a link to #1008 above

**Comment by AGBrown** [comment]
> should we perhaps pass the MathJax and image promises to hooks in an object, so they can call `.then()` on them if they want to?

I'd vote for this to be looked at. Currently you can't easily configure MathJax in a single place which limits its ease of use if you want common configuration for multiple note templates, or even on front and back cards in a single template. 

I tried loading a non-async js script with macro definitions from `collection.media`. However (I might be wrong, I did try a few different things) by the time it runs it looks like typesetting has already occurred, so calling `MathJax.startup.getComponents()` has no effect and the macros are not used. It is possible with a `\( \def ... \)` construct, but that has to be included on every card front and back template on every note type.

If it was possible to defer typesetting and give us a chance to load a js script from collection.media to hook into the [mathjax config events] to do extra config (e.g. to add macros, include extra packages such as `spalign` etc.) that would be a big win for using anki for mathematics.

[mathjax config events]: http://docs.mathjax.org/en/latest/web/configuration.html#performing-actions-during-startup

**Comment by hgiesel** [comment]
> Currently you can't easily configure MathJax in a single place which limits its ease of use if you want common configuration for multiple note templates.
Sharing assets across note types (or card types) is a common request, not just for MathJax, but also HTML/CSS/JS snippets. However it won't be trivial to implement.

Currently you can't easily configure MathJax in a single place which limits its ease of use if you want common configuration for multiple note templates

> I tried loading a non-async js script [...]

You can use [onUpdateHook](https://github.com/ankitects/anki/blob/76b005991e15d5d65229a1b227b17ffb7a182513/qt/aqt/data/web/js/reviewer.ts#L86) for this.
It takes a function that returns a Promise, which will be waited upon, so you can be ensure it executes before typesetting happens (just a few lines below).
IIRC, `onUpdateHook` does work on AnkiDroid, but not AnkiMobile.

**Comment by AGBrown** [comment]
@hgiesel apologies, I did not realise that existed. I'll try that out.

**Comment by hgiesel** [comment]
This concerns #1292 in its current state:

1. Is adressed (at least the opacity change)

You also mentioned swapping out / prerendering cards in this context. I disregarded shadow DOM, saying that `document.{querySelector,getElementById}`, etc. won't work. However this is not true, if we embed it in an `<iframe>`. The HTML of iframes can be set via the `srcdoc` attribute, e.g. `<iframe srcdoc="<html><head>...</head><body>..</body></html>" />`. This would make it perfect for our use case, because in #1292, in `convertMathJax` `html.document` is a full-fledged HTML document, which we could just transform to a string via `outerHTML`, and then set as srcdoc.

2. Again I don't really mind either approach, we should just be consistent. Even though I like unprefixed for its terseness, we should probably export into a prefix. If we go down this road, we should quickly change DeckOptions before 2.1.45, and adapt editor (while deprecating the old approach).

3. Honestly after thinking about this for a bit longer, my _ideal_ solution would be to completely deprecate JavaScript inside of templates, instead offer a notetype-wide "script section" in clayout, which allows for the definition of hook functions, which are called from `reviewer.ts` at appropriate times. One would have to be called before the cardtemplate is passed to MathJax (basically the old `onUpdate`), we could add one for modifying MathJax options, and one like `asShown`, which needs access to the `document` of the card.

Going a step further, `iframe`s allows for [sandboxing](https://www.w3schools.com/tags/att_iframe_sandbox.asp), with which we could even go as far as disabling `<script>`s within templates. This way, even if those new hooks I described fail, we still have a JS-less card we can safely display, and display a concrete message at the same time, like ("xxx hook failed"), which makes it obvious that it is due to the notetype settings, not Anki itself.

4. For MathJax, this is no longer applicable, for Images, see #3.
5. See #3.
6. Is addressed by #1292.

**Comment by dae** [comment]
An iframe could potentially help us when moving from one card to the next, but won't it break any templates that used globals or top level dom nodes to retain state between the question and the answer?

I've sometimes thought it would be nice if instead of having separate question and answer templates, there were just a single template, eg

```
{{Question}}
<hr>
<div id=answer>{{Answer}}</div>
```

The answer could be hidden by default, and 'show answer' would just toggle visibility, with optional css transitions. Users could potentially wrap other areas in .question to have it hidden when the answer is shown. And we could potentially allow overriding the whole behaviour with JS, so cards could 'reveal' in other ways if they wished.

But it wouldn't be compatible with what we currently have, so we'd need to introduce a per-notetype switch between the old and new behaviour. And it wouldn't work properly with video/sound embedded in the answer side without the autoplay logic being moved to JS.  And all clients would need to be updated to support it. It would be a lot of work :-) And I don't think we'll be in a good position to undertake that until the codebase is in better shape, and the clients are sharing more code, so we don't have to implement all the changes multiple times, or wait for clients to catch up.

Moving JS out of the templates and into something that can be shared across notetypes would be nice, and could potentially address a lot of the issues we currently have with stale/broken JS in templates from shared decks. It's something I'd like to see in the future, once the foundations have been improved some more. :-)

**Comment by dae** [comment]
Thanks for uploading that example add-on Henrik, I think that will make things a fair bit easier for people.

> Again I don't really mind either approach, we should just be consistent. Even though I like unprefixed for its terseness, we should probably export into a prefix. If we go down this road, we should quickly change DeckOptions before 2.1.45, and adapt editor (while deprecating the old approach).

Sorry, I didn't notice components was being exported as a global when I merged the PR. I agree - I think we should keep the globals to a minimum, and we'd better do it sooner rather than later. If practical, it would be great to keep everything under the 'anki' namespace (eg anki.components), so there's one place add-on/template authors know to look to see all exposed API surface - and it will minimize conflicts too. Would it make sense to put the promises like $editorToolbar in that namespace too?

Things like forEditorField that have already made it into a stable release aren't quite as urgent, as we'll need to keep aliases around for them anyway, but for new API it would be great to sort this out before a release.
